python/ray/train/v2/_internal/metrics/base.py

```diff
@@ -0,0 +1,165 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Dict, Generic, Optional, Tuple, TypeVar
+
+from ray.util.metrics import Gauge
+
+RUN_NAME_TAG_KEY = "ray_train_run_name"
+
+T = TypeVar("T")
+E = TypeVar("E", bound=Enum)
+
+
+class Metric(ABC):
+    def __init__(
+        self,
+        name: str,
+        default: T,
+        description: str,
+        base_tags: Dict[str, str],
+    ):
+        """
+        Initialize a new metric.
+
+        Args:
+            name: The name of the metric.
+            default: The default value of the metric.
+            description: The description of the metric.
+            base_tags: The base tags for the metric.
+        """
+        self._default = default
+        self._base_tags = base_tags
+        self._gauge = Gauge(
+            name,
+            description=description,
+            tag_keys=self._get_tag_keys(),
+        )
+
+    @abstractmethod
+    def record(self, value: T):
+        """Update the metric value.
+
+        Args:
+            value: The value to update the metric with.
+        """
+        pass
+
+    @abstractmethod
+    def get_value(self) -> T:
+        """Get the value of the metric.
+
+        Returns:
+            The value of the metric. If the metric has not been recorded,
+            the default value is returned.
+        """
+        pass
+
+    @abstractmethod
+    def reset(self):
+        """Reset values and clean up resources."""
+        pass
+
+    def _get_tag_keys(self) -> Tuple[str, ...]:
+        return tuple(self._base_tags.keys())
+
+
+class TimeMetric(Metric):
+    """A metric for tracking elapsed time."""
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        base_tags: Dict[str, str],
+    ):
+        self._current_value = 0.0
+        super().__init__(
+            name=name,
+            default=0.0,
+            description=description,
+            base_tags=base_tags,
+        )
+
+    def record(self, value: float):
+        """Update the time metric value by accumulating the time.
+
+        Args:
+            value: The time value to increment the metric by.
+        """
+        self._current_value += value
+        self._gauge.set(self._current_value, self._base_tags)
+
+    def get_value(self) -> float:
+        return self._current_value
+
+    def reset(self):
+        self._gauge.set(self._default, self._base_tags)
+        self._current_value = 0.0
```

Comment:
should we set it to default?

Reply:
or maybe just remove "Default" from the base class since it's not used there

Reply:
yes good call, I will remove since `Metric` is now abstract!

Reply:
Oh actually I will keep it and convert to use default because I think it's nice to guarantee this in the logic of `get_value`